src/util-lua-sandbox.c

```diff
@@ -0,0 +1,381 @@
+/* Copyright (C) 2023-2024 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Jo Johnson <pyrojoe314@gmail.com>
+ */
+
+#include "suricata-common.h"
+
+#include "lua.h"
+#include "lauxlib.h"
+#include "lualib.h"
+#include "util-debug.h"
+
+#include "util-debug.h"
+#include "util-validate.h"
+#include "util-lua-sandbox.h"
+
+#define SANDBOX_CTX "SANDBOX_CTX"
+
+static void HookFunc(lua_State *L, lua_Debug *ar);
+
+/**
+ * Lua allocator function provided to lua_newstate.
+ *
+ * \param ud The pointer passed to lua_newstate
+ * \param ptr Pointer to data being allocated/reallocated/freed
+ * \param osize Original size of the block
+ * \param nsize Size of the new block
+ *
+ * See: https://www.lua.org/manual/5.4/manual.html#lua_Alloc
+ */
+static void *LuaAlloc(void *ud, void *ptr, size_t osize, size_t nsize)
+{
+    (void)ud;
+    (void)osize;
+    SCLuaSbState *ctx = (SCLuaSbState *)ud;
+
+    if (nsize == 0) {
+        if (ptr == NULL) {
+            /* This happens, ignore. */
+            return NULL;
+        }
+        BUG_ON(osize > ctx->alloc_bytes);
+        SCFree(ptr);
+        ctx->alloc_bytes -= osize;
+        return NULL;
+    } else if (ptr == NULL) {
+        /* Allocating new data. */
+        void *nptr = SCRealloc(ptr, nsize);
+        if (nptr != NULL) {
+            ctx->alloc_bytes += nsize;
+        }
+        return nptr;
+    } else {
+        /* Resizing existing data. */
+        ssize_t diff = nsize - osize;
+
+        if (ctx->alloc_bytes + diff > ctx->alloc_limit) {
+            /* This request will exceed the allocation limit. Act as
+             * though allocation failed. */
+            ctx->memory_limit_error = true;
+            return NULL;
+        }
+
+        void *nptr = SCRealloc(ptr, nsize);
+        if (nptr != NULL) {
+            BUG_ON((ssize_t)ctx->alloc_bytes + diff < 0);
+            BUG_ON(osize > ctx->alloc_bytes);
```

Comment:
Before 8, should these become debug validates?

Reply:
yeah seems like a good idea